File: matmul.jl
# Import necessary library
using Random, Statistics, BenchmarkTools

# Set the random seed for reproducibility (optional)
Random.seed!(42)
#This is 10,000 times faster then the naive python version
# Create two random matrices (normal distribution, mean 0, sd 1)
A = randn(10, 784)  # 1000x1000 matrix with mean 0, sd 1
B = randn(784, 10)

# Time the multiplication of the two matrices

# Initialize the result matrix with zeros
C = zeros(10, 10)

# Time the multiplication of the two matrices using three nested loops
using BenchmarkTools

# In python this took 500 ms on my macbook 

@btime begin
      for i in 1:size(A, 1)  # Loop over the rows of A
          for j in 1:size(B, 2)  # Loop over the columns of B
             for k in 1:size(A, 2)  # Loop over the columns of A and rows of B
                C[i, j] += A[i, k] * B[k, j]  # Compute the dot product
            end
        end
    end
end
# result was 5.5 ms.  Numba did this same thing in 50 microsecods!!!  

@btime  C = A*B;
# result was 6 microseconds.  This is comparable to pytorch buildin (not on gpu) 

File: matmul.R
# Define matrix multiplication function using loops
matmul_basic <- function(A, B) {
  # Get dimensions of matrices
  n <- nrow(A)
  m <- ncol(B)
  p <- ncol(A)
  
  # Initialize result matrix with zeros
  C <- matrix(0, n, m)
  
  # Perform multiplication using three nested loops
  for (i in 1:n) {
    for (j in 1:m) {
      sum <- 0
      for (k in 1:p) {
        sum <- sum +  A[i, k] * B[k, j]
      }
      C[i,j] <- sum
    }
  }
  return(C)
}

# Test with random matrices
set.seed(123)
A <- matrix(rnorm(10 * 784), nrow = 10, ncol = 784)
B <- matrix(rnorm(784 * 10), nrow = 784, ncol = 10)

# Timing the loop-based matrix multiplication
system.time(C <- matmul_basic(A, B))*1000.0

# Compare with built-in matrix multiplication for reference
system.time(C_builtin <- A %*% B)*1000.0

# Load the package
library(microbenchmark)

# preallocate 
C <- matrix(0, 10,10)
C_builtin <- matrix(0,10,10)
# Use microbenchmark for more precise timing
microbenchmark(
  C = matmul_basic(A, B),
  C_builtin = A %*% B,
  times = 10
)

# result:  
#Unit: microseconds
#      expr      min       lq      mean    median       uq      max neval
#         C 2497.392 2610.183 2771.6205 2731.7275 2913.665 3131.621    10
# C_builtin   37.187   39.114   41.2993   40.5695   42.640   48.462    10